politeiawww/user/mysqldb/mysqldb.go

```diff
@@ -0,0 +1,1018 @@
+package mysqldb
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/binary"
+	"encoding/hex"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/decred/politeia/politeiawww/user"
+	"github.com/decred/politeia/util"
+	"github.com/google/uuid"
+	"github.com/marcopeereboom/sbox"
+
+	// MySQL driver.
+	_ "github.com/go-sql-driver/mysql"
+)
+
+const (
+	// Database options
+	connTimeout     = 1 * time.Minute
+	connMaxLifetime = 1 * time.Minute
+	maxOpenConns    = 0 // 0 is unlimited
+	maxIdleConns    = 100
+
+	// Database user (read/write access)
+	userPoliteiawww = "politeiawww"
+
+	databaseID = "users"
+
+	// Database table names.
+	tableNameKeyValue   = "key_value"
+	tableNameUsers      = "users"
+	tableNameIdentities = "identities"
+	tableNameSessions   = "sessions"
+
+	// Key-value store keys.
+	keyVersion             = "version"
+	keyPaywallAddressIndex = "paywalladdressindex"
+)
+
+// tableKeyValue defines the key-value table.
+const tableKeyValue = `
+  k VARCHAR(255) NOT NULL PRIMARY KEY,
+  v LONGBLOB NOT NULL
+`
+
+// tableUsers defines the users table.
+const tableUsers = `
+  ID VARCHAR(36) NOT NULL PRIMARY KEY,
+  username VARCHAR(64) NOT NULL,
+  uBlob BLOB NOT NULL,
+  createdAt INT(11) NOT NULL,
+  updatedAt INT(11),
+  UNIQUE (username)
+`
+
+// tableIdentities defines the identities table.
+const tableIdentities = `
+  publicKey CHAR(68) NOT NULL PRIMARY KEY,
+  userID    VARCHAR(36) NOT NULL,
+  activated INT(11) NOT NULL,
+  deactivated INT(11) NOT NULL,
+  FOREIGN KEY (userID) REFERENCES users(ID)
+`
+
+// tableSessions defines the sessions table.
+const tableSessions = `
+  k CHAR(64) NOT NULL PRIMARY KEY,
+  userID VARCHAR(36) NOT NULL,
+  createdAt INT(11) NOT NULL,
+  sBlob BLOB NOT NULL
+`
+
+var (
+	_ user.Database = (*mysqldb)(nil)
+)
+
+// mysqldb implements the user.Database interface.
+type mysqldb struct {
```

Comment:
This file needs unit tests.  The cockroachdb implementation uses `github.com/DATA-DOG/go-sqlmock`.  I don't have a strong opinion on sql mocking frameworks.  If there is another one you want to use I'm open to it.  Just run it by me first.  The `cockroachdb_test.go` tests can be used as an example, but I would prefer a cleaned up version of those.